Zend/Optimizer/block_pass.c

```diff
@@ -503,6 +503,19 @@ static void zend_optimize_block(zend_basic_block *block, zend_op_array *op_array
 								MAKE_NOP(opline);
 								++(*opt_count);
 								break;
+							case ZEND_TYPE_CHECK:
+								if (opline->opcode == ZEND_BOOL_NOT) {
+									if (src->extended_value == MAY_BE_RESOURCE || src->extended_value == (MAY_BE_ANY - MAY_BE_RESOURCE)) {
```

Comment:
I'm not sure this holds.

https://github.com/php/php-src/blob/6422cf6f1a427977812187a2fb2dee26359b5884/Zend/zend_compile.c#L3950

For `is_scalar` we use `ZEND_TYPE_CHECK`. Negating this mask would result in:

```c
MAY_BE_NULL|MAY_BE_STRING|MAY_BE_ARRAY|MAY_BE_OBJECT|MAY_BE_RESOURCE
```

However, here we now contain `MAY_BE_RESOURCE` which means for `!is_scalar($resource)` we'd get `false` when the `resource` is closed. Unfortunately I don't think we can avoid the `ZEND_BOOL_NOT` here.

Reply:
Ah, `ZEND_TYPE_CHECK` is implemented in such a way that closed resources are only considered not to be resources when `opline->extended_value == MAY_BE_RESOURCE`. In that case this should be fine.